### PR TITLE
ESM_FORMAT: Avoid reading package.json when extension is unambiguous

### PR DESCRIPTION
**Pull request by mathiasbynens**
It’s wasteful to load the package.json even if the extension is unambiguous. This patch optimizes for the zero-config case by avoiding needless package.json loads.

With this patch, the mental model becomes a lot simpler too:

- `.cjs` means `commonjs`
- `.mjs` means `module`

See discussion thread: https://twitter.com/mathias/status/1096167102455709696

##### Checklist

- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

**Comment by ljharb** [comment]
I don’t think this is a future-looking optimization; eventually package.json should be able to override any extension to parse in any way.

**Comment by mathiasbynens** [comment]
That’s what’s under discussion. IMHO it’s worth [carving out `.cjs` and `.mjs`](https://twitter.com/mathias/status/1096170719195660289) since they’re unambiguous, and only enable overriding other extensions. From the Twitter thread it sounded like people were on board 

We need to weigh the cost of allowing `.cjs` and `.mjs` to be overridden (i.e. load `package.json` all the time) with the performance benefits of not having to do so for the zero-config case.

**Comment by robpalme** [comment]
This PR looks good.

My view is that package.json is about defining package-scoped things, whereas extensions/pragmas are a way of specialising/overriding behaviour on a per-file basis.

This locality simplifies the mental model and eliminates a data-dependency.

**Comment by guybedford** [comment]
Thanks @mathiasbynens for the PR. This is a simplification of the existing spec that seems sensible to me. 

There is effectively only one behavioural change here from what was written previously and that is that previously if there was no package boundary, we would load `.mjs` files as CommonJS which seems wrong anyway.

This sounds like a spec bug to me, where the fix happens to enable an optimization.

We as a group can decide over time whether we want that optimization to remain, but what we're really deciding from a behavioural perspective is if `node x.mjs` where there is no package.json at all should load as a module.

**Review by guybedford — APPROVED** [review]

**Comment by mathiasbynens** [comment]
> what we're really deciding from a behavioural perspective is if `node x.mjs` where there is no `package.json` at all should load as a module.

+1 FWIW, there is precedent for this in both `jsc` and `d8`. It seems reasonable, and even expected, to me.

**Comment by guybedford** [comment]
Note that the implementation already does this too, so that this is actually just a spec bug.

**Comment by GeoffreyBooth** [comment]
This is fine for where the implementation is now, but if we add package exports we'll always need to read package.json to know what the valid paths inside a package are. Also package-level loaders, if added, would also require always reading package.json. Ditto for file extension configuration like @ljharb said.

Another point is if `type: module` ever defines anything in a package more than just file extension mappings, separately than what .mjs defines (globals like `Buffer` come to mind, and the discussion about making them package-scoped).

But for right now, sure.

**Review by robpalme — APPROVED** [review]

File: doc/api/esm.md [hunk 1]
```diff
@@ -251,24 +251,19 @@ PACKAGE_MAIN_RESOLVE(_packageURL_, _pjson_)
...
->    1. If _url_ ends in _".cjs"_, then
->       1. Return _"commonjs"_.
 >    1. If _url_ does not end in _".js"_ or _".mjs"_, then
 >       1. Throw an _Unsupported File Extension_ error.
 >    1. Return _"module"_.
-> 1. Otherwise,
->    1. If _url_ ends in _".mjs"_, then
->       1. Return _"module"_.
->    1. Otherwise,
->       1. Return _"commonjs"_.
```
Comment (guybedford) [inline doc/api/esm.md]: Note this line needs to remain though - 

```
> 1. Otherwise,
>   1. Return _"commonjs"_.
```

**Review by guybedford — CHANGES_REQUESTED** [review]
Just noticed this line is missing.

File: doc/api/esm.md (hunk 1 above)
Comment (mathiasbynens) [reply doc/api/esm.md]: Done!

**Review by guybedford — APPROVED** [review]

**Review by MylesBorins — APPROVED** [review]
Lgtm

**Review by ljharb — CHANGES_REQUESTED** [review]
We don’t even have consensus that the cjs  extension will be a thing, and no extension should be special - if any of them can’t br overridden, then none of them can (including js)

**Comment by guybedford** [comment]
@ljharb please read https://github.com/nodejs/ecmascript-modules/pull/33#issuecomment-464334948 and then let me know if you want to revise your comment.

**Comment by ljharb** [comment]
@guybedford that comment hasn’t made anything clear to me about why I’d revise it.

In the absence of a package.json, and perhaps when run as an entry point, it makes sense that a file extension unambiguously determines the parse goal, but if a package.json is present (and certainly when it’s not an entry point), it must always be consulted regardless of the file extension. No extension should be special; either they can all be overridden from package.json or none of them can.

Separately, adding the cjs extension, without any consensus besides apparent side channel discussions about it, isn’t something that should be done via a PR.

**Comment by guybedford** [comment]
> In the absence of a package.json, and perhaps when run as an entry point, it makes sense that a file extension unambiguously determines the parse goal

Currently, by the spec, `node x.mjs` where there is no package.json will execute as CommonJS, so this is not the case as written.

The change here ONLY changes the above. That is the only behavioural change.

**Comment by ljharb** [comment]
Then if the cjs extension is removed from the PR, it seems like my objection will be unwarranted. Thanks for clarifying.

**Comment by ljharb** [comment]
(and if cjs is already in this repo, then that’s not something that ever got consensus, so it doesn’t need to block this PR but it does need to be removed ASAP)

**Comment by GeoffreyBooth** [comment]
> Separately, adding the cjs extension, without any consensus besides apparent side channel discussions about it, isn’t something that should be done via a PR.

On 2019-01-30 I opened [a PR](https://github.com/nodejs/modules/pull/251) to add handling initial entry points to our Phase 2 roadmap, and I invited anyone in the group to collaborate with me in putting together a proposal. At the time I didn’t have a specific implementation in mind (no new file extensions or flags were yet proposed). A few members of the group _did_ contact me, and we had many side discussions which led to a proposal.

On 2019-02-06 I opened [an issue](https://github.com/nodejs/modules/issues/262) with a link to [the proposal](https://github.com/GeoffreyBooth/node-esm-entry-points-proposal) that the small group produced. That proposal very clearly proposed a `.cjs` file extension, with a section devoted to the topic: https://github.com/GeoffreyBooth/node-esm-entry-points-proposal#cjs-file-extension. Many members of the group opened issues on that proposal to offer feedback, and we made changes.

On 2019-02-10 @guybedford opened [a PR](https://github.com/nodejs/ecmascript-modules/pull/32) on this repo to implement most of the above proposal (there were a few parts he didn’t have time to get to, that we plan to do in weeks ahead). The first two lines of the post introducing the PR read:

> This implements a top-level `--type` flag for `--experimental-modules` with the following behaviours.
> 
> In addition `"type": "esm"` in the package.json is renamed to `"type": "module"`, and a new extension `.cjs` is supported.

On 2019-02-13 we had a meeting where @MylesBorins brought up that PR. [He read those first two lines aloud](https://youtu.be/W25K1c2e2y8?t=3164) and asked for objections, and no one objected.

So `.cjs` was created with consensus. At this point a new proposal would need to be offered to remove it from this implementation.

@ljharb I understand you strongly desire a file extensions map, so why don’t you propose it to the group? You can follow a process like the one I just did for creating `.cjs`.

File: doc/api/esm.md [hunk 2]
```diff
@@ -251,24 +251,21 @@ PACKAGE_MAIN_RESOLVE(_packageURL_, _pjson_)
...
+>    1. If _isMain_ is **true** and the `--type` flag is _"module"_, then
+>       1. Throw an _Invalid File Extension_ error.
+>    1. Return _"commonjs"_.
 > 1. Let _pjson_ be the result of **READ_PACKAGE_BOUNDARY**(_url_).
 > 1. If _pjson_ is **null** and _isMain_ is **true**, then
 >    1. Return _"commonjs"_.
 > 1. If _pjson.type_ exists and is _"module"_, then
->    1. If _url_ ends in _".cjs"_, then
->       1. Return _"commonjs"_.
 >    1. If _url_ does not end in _".js"_ or _".mjs"_, then
```
Comment (ChALkeR) [inline doc/api/esm.md]:
```suggestion
>    1. If _url_ does not end in _".js"_, then
```

`.mjs` case is non-reachable as already processed above.

**Comment by MylesBorins** [comment]
@ljharb if we want to have a file extension map won't we be able to extend this proposal at that time? One thought I'm having is that with "loaders" we will potentially want to hook in before our resolution happens at all. With that in mind potential extensions to this algorithm, such as extension mapping, could be treated as a custom built in loader if they exist.

As such... isn't this fine to land based on the current implementation and we can iterate as we figure out the implementation for a map?

**Comment by devsnek** [comment]
i'd object to `mode: esm` ever being unflagged because it simultaneously leaves out what it is mapping (.js) and what the result of that mapping is (source text module)

**Comment by GeoffreyBooth** [comment]
> edit: sorry i misunderstood what myles was talking about. this pr itself is fine (still shouldn’t be unflagged though)

Thanks @devsnek. I think it might be useful to have a place somewhere (Google Doc? Issue?) where we consolidate each member’s general objections to the new implementation, and how severe they are. Like you want feature X added, and you’re willing to let the new implementation ship flagged without it but you’re not willing to drop the `--experimental-modules` flag unless X is added, etc. It’s hard to keep track of everyone’s broader objections when they’re lodged as requests for changes to individual PRs.

**Comment by Trott** [comment]
I imagine this can/should be closed at this point? Or if not closed, then at least transferred to the main repository?

**Pull request closed without merging**